README.md
Wormhole
=========
[![Build Status on TravisCI](https://secure.travis-ci.org/kiesel/wormhole-go.png)](http://travis-ci.org/kiesel/wormhole-go)
[![GitHub release](https://img.shields.io/github/release/kiesel/wormhole-go.svg?maxAge=2592000)](https://github.com/kiesel/wormhole-go/releases)
[![license](https://img.shields.io/github/license/kiesel/wormhole-go.svg?maxAge=2592000)](https://github.com/kiesel/wormhole-go/blob/master/LICENSE)

Wormhole is a an application that allows to open files from a commandline within a VM in your favorite editor(s) / applications in the host system.

Prerequisites are that:

* the part of the VM filesystem that is hosting the files in question is mounted in your host OS
* you are logging in via SSH (though that limitation is only relevant for the client part.)

Installation
------------

1. Download the latest release from the [GitHub Releases](https://github.com/kiesel/wormhole-go/releases) page.
2. Extract `.wormhole.yml` from the release zip into your home directory.
3. Run `wormhole` / `wormhole.exe`

To start wormhole with your shell, put this line into `.bashrc` / `.zshrc`:

```sh
(nohup $HOME/wormhole.exe -quiet 2>&1 &)
```

Client installation
-------------------
You'll need a client, too. Given you're using bash / zsh or a compatible shell, you can use [kiesel/wormhole](https://github.com/kiesel/wormhole).


Security
--------

Wormhole opens a port on a designated interface for you; when binding to public network interfaces, you might expose yourself to serious security risks.

The recommended setup is therefore, to bind it to the loopback address 127.0.0.1 / ::1 and use SSH to make that port available to your client system (which would usually be a local VM). The configuration file `.ssh/config` would look like this:

    Host vbox vb 127.0.0.1
      Hostname 127.0.0.1
      Port 2222

      RemoteForward 127.0.0.1:5115 127.0.0.1:5115


ChangeLog.md
Wormhole GO
===========

2016-04-19 - Release 1.1.1
--------------------------
* Fixed argument merging for apps (@kiesel)

2016-04-19 - Release 1.1.0
--------------------------
* Improved support for program invocations with spaces
  in the path (@kiesel)
* Improved test suite (@kiesel)


2016-04-15 - Release 1.0.0
---------------------------
* Released first stable version: 1.0.0 (@kiesel)
* Supported commands:
  - invoke
  - reload
  - version
  - exit



lib/config_test.go
package wormhole

import (
	"reflect"
	"testing"
)

func deepEqual(expect, actual interface{}, t *testing.T) {
	if !reflect.DeepEqual(expect, actual) {
		t.Error("Items not equal:", expect, actual)
	}
}

func TestDefaultAddress(t *testing.T) {
	config := &WormholeConfig{}
	if "127.0.0.1:5115" != config.GetAddr() {
		t.Error("Wrong default address: " + config.GetAddr())
	}
}

func TestReadSimpleConfig(t *testing.T) {
	config, err := ReadConfiguration([]byte(`
addr: :5115

mapping:
  "/home/": "A:"

apps:
  "sublime": "/opt/sublime/sublime"
  `))

	if err != nil {
		t.Error("Failure during parsing", err.Error())
	}

	deepEqual(map[string]string{"/home/": "A:"}, config.Mapping, t)

	var app *App
	if app, err = config.GetApp("sublime"); err != nil {
		t.Error(err)
	}

	deepEqual("/opt/sublime/sublime", app.Executable, t)
}

func Test_read_config_unquoted_with_args(t *testing.T) {
	createConfigAndCompareApp(`
apps:
  app: cmd.exe /c start
  `, &App{
		Executable: "cmd.exe /c start",
		Args:       []string{},
	},
		t)
}

func Test_read_config_quoted(t *testing.T) {
	createConfigAndCompareApp(`
apps:
  app: "C:/Documents and Settings/User/kiesel/cmd.exe"
  `, &App{
		Executable: "C:/Documents and Settings/User/kiesel/cmd.exe",
		Args:       []string{},
	},
		t)
}

func Test_read_config_array_notation(t *testing.T) {
	createConfigAndCompareApp(`
apps:
  app: ["cmd.exe", "/c", "start"]
  `, &App{
		Executable: "cmd.exe",
		Args:       []string{"/c", "start"},
	},
		t)
}

func Test_read_config_unquoted_with_whitespace_path_leads_to_broken_executable(t *testing.T) {
	createConfigAndCompareApp(`
apps:
  app: cmd with whitespace.exe /c start
  `, &App{
		Executable: "cmd with whitespace.exe /c start",
		Args:       []string{},
	},
		t)
}

func Test_read_config_quoted_with_args_in_array_notation(t *testing.T) {
	createConfigAndCompareApp(`
apps:
  app: ["cmd with whitespace.exe", "/c", "start"]
  `, &App{
		Executable: "cmd with whitespace.exe",
		Args:       []string{"/c", "start"},
	},
		t)
}

func Test_read_config_in_array_notation_then_merge_with_args(t *testing.T) {
	config, err := ReadConfiguration([]byte(`
apps:
  app: ["cmd with whitespace.exe", "/c", "start"]
  `))

	if err != nil {
		t.Error("Failure parsing", err.Error())
	}

	var app *App
	if app, err = config.GetAppWith("app", []string{"foo", "bar"}); err != nil {
		t.Error(err)
	}

	deepEqual([]string{"/c", "start", "foo", "bar"}, app.Args, t)
}

func createConfigAndCompareApp(cstr string, expect *App, t *testing.T) *App {
	config, err := ReadConfiguration([]byte(cstr))

	if err != nil {
		t.Error("Failure during parsing", err.Error())
		return nil
	}

	var app *App
	if app, err = config.GetApp("app"); err != nil {
		t.Error("Failure to get app", err)
		return nil
	}

	deepEqual(expect, app, t)
	return app
}


lib/config.go
package wormhole

import (
	"errors"
	"io/ioutil"
	"os"
	"path"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v2"
)

type Error interface {
	Error() string
}

type WormholeConfig struct {
	Addr    string            `yaml:"listen,omitempty"`
	Mapping map[string]string `yaml:"mapping"`
	App     map[string]App    `yaml:"apps"`
}

func GetDefaultConfig() string {
	return path.Join(os.Getenv("HOME"), ".wormhole.yml")
}

func GetDefaultLog() string {
	return path.Join(os.Getenv("HOME"), "wormhole.log")
}

func ReadConfigurationFrom(filename string) (config *WormholeConfig, err Error) {
	source, err := ioutil.ReadFile(filename)
	if err != nil {
		log.Critical(err.Error())

		return nil, err
	}

	return ReadConfiguration(source)
}

func ReadConfiguration(source []byte) (config *WormholeConfig, err Error) {
	err = yaml.Unmarshal(source, &config)
	if err != nil {
		log.Critical(err.Error())
		return nil, err
	}

	return config, nil
}

func (this *WormholeConfig) GetAddr() string {
	if "" == this.Addr {
		return "127.0.0.1:5115"
	}

	return this.Addr
}

func (this *WormholeConfig) GetApp(key string) (*App, Error) {
	app, ok := this.App[key]

	if !ok {
		return nil, errors.New("No mapping for '" + key + "'")
	}

	return &app, nil
}

func (this *WormholeConfig) GetAppWith(key string, args []string) (app *App, err Error) {
	if app, err = this.GetApp(key); err != nil {
		return app, err
	}

	app.MergeArguments(this.translatePaths(args))
	return app, nil
}

func (this *WormholeConfig) AvailableApps() string {
	var keys []string

	for key, _ := range this.App {
		keys = append(keys, key)
	}

	return strings.Join(keys, ", ")
}

func (this *WormholeConfig) translatePath(path string) string {
	for from, to := range this.Mapping {
		path = strings.Replace(path, from, to, 1)
	}

	return filepath.FromSlash(path)
}

func (this *WormholeConfig) translatePaths(paths []string) []string {
	for index, arg := range paths {
		paths[index] = this.translatePath(arg)
	}

	return paths
}

type App struct {
	Executable string
	Args       []string
}

func (this *App) MergeArguments(args []string) {
	this.Args = append(this.Args, args...)
}

// Callback for YAML unmarshalling
func (this *App) UnmarshalYAML(unmarshal func(interface{}) error) (err error) {

	// First attempt: just treat value as a single string - that is executable w/o any
	// arguments
	var line string
	if err = unmarshal(&line); err == nil {

		// Try to split string into executable and optional args...
		this.Executable = strings.TrimSpace(line)
		this.Args = []string{}
		return err
	}

	// Attempt: treat as array value, first being the executable all others arguments
	var array []string
	if err = unmarshal(&array); err == nil {
		this.Executable = array[0]
		this.Args = array[1:]
	}

	return err
}


lib/execute.go
package wormhole

import (
	"gopkg.in/op/go-logging.v1"
	"io"
	"os/exec"
)

var log = logging.MustGetLogger("wormhole")

func transcribeOutput(prefix string, stream io.ReadCloser) {
	var buf = make([]byte, 1024)

	for {
		n, err := stream.Read(buf)

		if n > 0 {
			log.Info("%s: %s", prefix, buf)
		}

		if err != nil {
			return
		}
	}
}

func ExecuteCommand(executable string, args ...string) (err Error) {
	cmd := exec.Command(executable, args...)

	stdout, err := cmd.StdoutPipe()
	if err != nil {
		log.Error(err.Error())
		return err
	}
	defer stdout.Close()

	stderr, err := cmd.StderrPipe()
	if err != nil {
		log.Error(err.Error())
		return err
	}
	defer stderr.Close()

	if err := cmd.Start(); err != nil {
		log.Error(err.Error())
		return err
	}

	log.Info("Started '%s' w/ PID %d", executable, cmd.Process.Pid)

	go transcribeOutput("out", stdout)
	go transcribeOutput("err", stderr)
	cmd.Wait()

	log.Info("PID %d has quit.", cmd.Process.Pid)

	return nil
}


wormhole/wormhole.go
package main

import (
	"bufio"
	"errors"
	"flag"
	"fmt"
	"net"
	"os"
	"strings"

	"github.com/kiesel/wormhole-go/lib"
	"gopkg.in/op/go-logging.v1"
)

type Error interface {
	Error() string
}

var (
	// Compile time values
	VersionString string

	// Command line flags
	quiet          bool
	displayVersion bool
	configFilename string
	logFilename    string

	// Volatiles
	config wormhole.WormholeConfig
	log    = logging.MustGetLogger("wormhole")
	format = logging.MustStringFormatter(
		"%{color}%{time:15:04:05.000} %{level:.1s} %{id:03x}%{color:reset} >> %{message}",
	)
)

func init() {
	flag.BoolVar(&displayVersion, "version", false, "Show version number, then exit.")
	flag.BoolVar(&quiet, "quiet", false, "Enable quiet mode")
	flag.StringVar(&configFilename, "configfile", wormhole.GetDefaultConfig(), "Set configuration path")
	flag.StringVar(&logFilename, "log", wormhole.GetDefaultLog(), "Set log path")
}

func Version() string {
	if "" != VersionString {
		return VersionString
	}

	return "<snapshot>"
}

func readConfiguration() (err Error) {
	var newConfig *wormhole.WormholeConfig

	log.Info("Trying to parse wormhole configuration from " + configFilename)
	if newConfig, err = wormhole.ReadConfigurationFrom(configFilename); err != nil {
		return err
	}

	// Now replace existing config with new
	log.Debug("New configuration %v", newConfig)
	config = *newConfig

	return nil
}

func main() {
	flag.Parse()

	if displayVersion {
		fmt.Println("Wormhole Version " + Version())
		os.Exit(1)
	}

	// Setup logging
	var logbackend *logging.LogBackend

	if quiet {
		fmt.Println("Quiet mode, logging to " + logFilename)

		logfile, err := os.OpenFile(logFilename, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0600)
		if err != nil {
			panic(err)
		}

		logbackend = logging.NewLogBackend(logfile, "", 0)
	} else {
		logbackend = logging.NewLogBackend(os.Stdout, "", 0)
	}

	logbackendformatter := logging.NewBackendFormatter(logbackend, format)
	logging.SetBackend(logbackendformatter)

	// Read config
	readConfiguration()

	// Start main
	log.Info("Wormhole %s server starting ...", Version())

	l, err := net.Listen("tcp4", config.GetAddr())
	if err != nil {
		log.Fatal(err)
	}

	log.Info("Listening at " + l.Addr().String())

	defer l.Close()
	for {
		// Wait for connection
		conn, err := l.Accept()

		if err != nil {
			log.Critical(err.Error())
			continue
		}

		// Handle connection
		go handleConnection(conn)
	}
}

func handleConnection(c net.Conn) {
	log.Debug("Received connection from %s", c.RemoteAddr().String())

	defer c.Close()
	defer log.Debug("Closed connection to %s", c.RemoteAddr().String())

	line, err := bufio.NewReader(c).ReadString('\n')
	if err != nil {
		log.Critical(err.Error())
		return
	}

	writer := bufio.NewWriter(c)

	log.Debug("[%s] >> %s", c.RemoteAddr().String(), line)
	resp, err := handleLine(c, line)

	if err != nil {
		log.Warning("[%s] << %s", c.RemoteAddr().String(), err.Error())

		writer.WriteString("[ERR] ")
		writer.WriteString(err.Error())
	} else {
		log.Info("[%s] << %s", c.RemoteAddr().String(), resp)

		writer.WriteString("[OK] ")
		writer.WriteString(resp)
	}

	writer.WriteString("\n")
	writer.Flush()
}

func handleLine(c net.Conn, line string) (resp string, err Error) {
	parts := strings.Split(strings.TrimSpace(line), " ")

	log.Debug("Extracted parts %s", parts)
	if len(parts) < 1 {
		log.Warning("Too few parts.")
		return "", errors.New("Too few words, expected at least 1.")
	}

	switch strings.ToLower(parts[0]) {
	case "invoke":
		if len(parts) < 2 {
			log.Warning("Too few parts.")
			return "", errors.New("Too few words, expected at least 2.")
		}

		return handleInvocation(parts[1], parts[2:])
	case "version":
		return handleVersion()
	case "exit":
		return handleExit()
	case "reload":
		return handleReload()
	}

	return "", errors.New("Unknown command, expected one of " + config.AvailableApps())
}

func handleInvocation(mapping string, args []string) (resp string, err Error) {
	app, err := config.GetAppWith(mapping, args)
	if err != nil {
		return "", err
	}

	log.Info("Invoking '%v' (mapped by %v) with args: %v", app.Executable, mapping, app.Args)
	go wormhole.ExecuteCommand(app.Executable, app.Args...)

	return "Started " + mapping, nil
}

func handleExit() (response string, err Error) {
	log.Warning("Client requested exit, exiting.")

	os.Exit(0)
	return "Bye!", nil
}

func handleReload() (response string, err Error) {
	if err := readConfiguration(); err != nil {
		return "", err
	}

	return "Re-read configuration.", nil
}

func handleVersion() (response string, err Error) {
	return Version(), nil
}
